src/client/keycode.cpp

```diff
@@ -221,122 +223,153 @@ static const struct table_key table[] = {
+}
 
-	// It's not a known key, complain and try to do something
-	Key = irr::KEY_KEY_CODES_COUNT;
-	int chars_read = mbtowc(&Char, name, 1);
-	FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-	m_name = "";
-	warningstream << "KeyPress: Unknown key '" << name
-		<< "', falling back to first char." << std::endl;
+KeyPress::KeyPress(std::string_view name)
+{
+	if (loadFromScancode(name))
+		return;
+	const auto &key = lookup_keyname(name);
+	Keycode keycode(key.Key, key.Char);
+	scancode = RenderingEngine::get_raw_device()->getScancodeFromKey(keycode);
 }
 
-KeyPress::KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character)
+KeyPress::KeyPress(const irr::SEvent::SKeyInput &in)
 {
-	if (prefer_character)
-		Key = irr::KEY_KEY_CODES_COUNT;
+#if USE_SDL2
+	if (in.SystemKeyCode)
+		scancode.emplace<u32>(in.SystemKeyCode);
 	else
-		Key = in.Key;
-	Char = in.Char;
-
-	try {
-		if (valid_kcode(Key))
-			m_name = lookup_keykey(Key).Name;
-		else
-			m_name = lookup_keychar(Char).Name;
-	} catch (UnknownKeycode &e) {
-		m_name.clear();
-	};
+		scancode.emplace<irr::EKEY_CODE>(in.Key);
+#else
+	Keycode keycode(in.Key, in.Char);
+	scancode = RenderingEngine::get_raw_device()->getScancodeFromKey(keycode);
+#endif
 }
 
-const char *KeyPress::sym() const
+std::string KeyPress::formatScancode() const
 {
-	return m_name.c_str();
+#if USE_SDL2
+	if (auto pv = std::get_if<u32>(&scancode))
+		return *pv == 0 ? "" : "<" + std::to_string(*pv) + ">";
+#endif
+	return "";
 }
 
-const char *KeyPress::name() const
+std::string KeyPress::sym() const
 {
-	if (m_name.empty())
-		return "";
-	const char *ret;
-	if (valid_kcode(Key))
-		ret = lookup_keykey(Key).LangName;
-	else
-		ret = lookup_keychar(Char).LangName;
-	return ret ? ret : "<Unnamed key>";
+	std::string name = lookup_scancode(scancode).Name;
+	if (USE_SDL2 || name.empty())
+		if (auto newname = formatScancode(); !newname.empty())
+			return newname;
+	return name;
 }
 
-const KeyPress EscapeKey("KEY_ESCAPE");
+std::string KeyPress::name() const
+{
+	const auto &name = lookup_scancode(scancode).LangName;
+	if (!name.empty())
+		return name;
+	return formatScancode();
+}
+
+irr::EKEY_CODE KeyPress::getKeycode() const
+{
+	return lookup_scancode(scancode).Key;
+}
+
+wchar_t KeyPress::getKeychar() const
+{
+	return lookup_scancode(scancode).Char;
+}
 
-const KeyPress LMBKey("KEY_LBUTTON");
-const KeyPress MMBKey("KEY_MBUTTON");
-const KeyPress RMBKey("KEY_RBUTTON");
+bool KeyPress::loadFromScancode(std::string_view name)
+{
+#if USE_SDL2
+	if (name.size() < 2 || name[0] != '<' || name.back() != '>')
+		return false;
+	char *p;
+	const auto code = strtoul(name.data()+1, &p, 10);
+	if (p != name.data() + name.size() - 1)
+		return false;
+	scancode.emplace<u32>(code);
+	return true;
+#else
+	return false;
+#endif
+}
+
+std::unordered_map<std::string, KeyPress> KeyPress::specialKeyCache;
+const KeyPress &KeyPress::getSpecialKey(const std::string &name)
```

Comment:
string_view

Reply:
We need a `std::string` in the implementation for indexing `specialKeyCache` so I don't think using `string_view` here would bring much.